Synchronize variable reification against real class.

headius opened the pull request:
This is likely at least a partial fix for #5910, where it appears
that multiple threads all trigger reificatin of a class's
variables at the same time, leading to some of those threads
choosing a different reified type and ultimately class casting
exceptions.

The change here synchronizes against the class's "real" class,
which should be the one on which we attach the allocated. This
should force multiple threads all allocating the first instance
at once to wait for each other, eventually falling back on the
new allocator that does not re-reify the type.

This will eventually be moot once individual object instances hold
a reference to their own layout managers.

headius commented:
@enebo @kares I have concerns about the locking.

kares commented:
looks reasonable (given no further details #5910 - still weird that there's only one pathological case).

would it make sense to reduce the lock duration to the bare minimum of (double-checking) and setting the generated class-and-allocator?, moving 'only' these parts under the critical section: https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/specialized/RubyObjectSpecializer.java#L152-L153

rdubya commented on core/src/main/java/org/jruby/RubyObject.java:
```diff
@@ -152,18 +153,28 @@ public IRubyObject allocate(Ruby runtime, RubyClass klass) {
+                } else {
+                    // someone else replaced allocator while we waited for lock, use it
+                    return allocator.allocate(runtime, klass);
+                }
```
Maybe it would affect readability, but would it be possible to get rid of the `else` and always call `return allocator.allocate(runtime, klass);` at the bottom? Not sure if it would have any benefits over just not having duplicated code though (if it is even possible.)

headius replied:
Agreed.

enebo commented:
I have been pondering impact of this and I believe the ordinary classes will not be much of an impact because there are just not that many types in a typical application.   A few thousand locks when most happen early in an app does not seem like a big deal.

Singleton types perhaps in some DCI pattern?  Two threads singletonizing the same object at the same time already hurts my head.  We must already synch on that?   Trying to think of patterns where a framework would endlessly make new types.  Seems like the main worry to more locking here.

headius commented:
> would it make sense to reduce the lock duration to the bare minimum of (double-checking) and setting the generated class-and-allocator?

The only thing that could move out would be the walking of all classes and methods in the type's hierarchy. Seems like that should be ok; the structures involved are all safe.

headius commented:
It turns out that the related Ruby code in #5910 was actually buggy to begin with; it allowed uninitialized classes to be constructed, which led to us seeing multiple views of instance variables in defined methods. I've detailed that here: https://github.com/jruby/jruby/issues/5910#issuecomment-543429744

We can't fix that, but it did expose some weaknesses in our variable reification that I have fixed in the most recent commit:

* We were forcing all methods to compile, breaking the startup-time promises of our lazy IRMethod compilation. That is now fixed; if the `defNode` is still present, we will walk it rather than forcing a compile. @enebo please review. We should investigate whether we're forcing methods to compile too early for other things.
* The walk of methods should produce the same result for the same unchanging method hierarchy. It did not only because the hierarchy was actually changing in #5910. Because the walk should produce the same result every time, I've moved it outside the synchronization block to reduce contention (but possibly waste some work). @kares please review.
